In [1]:
import numpy as np

SplitLine = "\n==============================\n"

def PrintArray(info:str, x):
    print(info + ":\n", x)
    print(f"维度数(秩)：{x.ndim}; 元素类型：{x.dtype}; 数组的维度形状：{x.shape}; 元素个数：{x.size}; 元素的字节大小：{x.itemsize}", end=SplitLine)

|函数|描述|
|:---|:---|
|dot|两个数组的点积|
|vdot|两个向量的点积|
|inner|两个数组的内积|
|matmul|两个数组的矩阵积|
|determinant|数组的行列式|
|solve|求解线性矩阵方程|
|inv|计算矩阵的乘法逆矩阵|

* numpy.dot() 两个数组的矩阵乘积        
对于多维数组，它的通用计算公式如下，即结果数组中的每个元素都是：数组a的最后一维上的所有元素与数组b的倒数第二位上的所有元素的乘积和

In [8]:
a = np.arange(1, 7).reshape((2, 3))
b = np.arange(5, 11).reshape((3, 2))
PrintArray("a", a)
PrintArray("b", b)

a:
 [[1 2 3]
 [4 5 6]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 3); 元素个数：6; 元素的字节大小：8
b:
 [[ 5  6]
 [ 7  8]
 [ 9 10]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 2); 元素个数：6; 元素的字节大小：8


In [10]:
c = np.dot(a, b)
PrintArray("c", c)

c_ = a.dot(b)
# c_ = a * b 为对应元素相乘，需要shape相同或者广播后可支持操作
PrintArray("c_", c_)

c:
 [[ 46  52]
 [109 124]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
c_:
 [[ 46  52]
 [109 124]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8


* numpy.vdot() 函数是两个向量的点积。 如果第一个参数是复数，那么它的共轭复数会用于计算。 如果参数是多维数组，它会被展开。

In [11]:
c = np.vdot(a, b)
PrintArray("c", c)

c:
 175
维度数(秩)：0; 元素类型：int64; 数组的维度形状：(); 元素个数：1; 元素的字节大小：8


* numpy.inner() 函数返回一维数组的向量内积。对于更高的维度，它返回最后一个轴上的和的乘积。

In [24]:
PrintArray("a", a)
PrintArray("b", b)
b_t = b.T
PrintArray("b_t", b_t)


c = np.inner(a, b_t)
PrintArray("c", c)

c_ = np.dot(a, b)
PrintArray("c_", c_)

print(np.array_equal(c, c_))

a:
 [[1 2 3]
 [4 5 6]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 3); 元素个数：6; 元素的字节大小：8
b:
 [[ 5  6]
 [ 7  8]
 [ 9 10]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 2); 元素个数：6; 元素的字节大小：8
b_t:
 [[ 5  7  9]
 [ 6  8 10]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 3); 元素个数：6; 元素的字节大小：8
c:
 [[ 46  52]
 [109 124]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
c_:
 [[ 46  52]
 [109 124]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
True


计算方式    [[1*5+2*7+3*9=46, 1*6+2*8+3*10=52], [4*5+5*7+6*9=109, 4*6+5*8+6*10=124]]

* numpy.matmul 函数返回两个数组的矩阵乘积。 虽然它返回二维数组的正常乘积，但如果任一参数的维数大于2，则将其视为存在于最后两个索引的矩阵的栈，并进行相应广播。

In [34]:
PrintArray("a", a)
PrintArray("b", b)

# dim=2
c = np.matmul(a, b)
PrintArray("c", c)

c_dot = a.dot(b)
PrintArray("c_dot", c_dot)
c_inner = np.inner(a, b.T)
PrintArray("c_inner", c_inner)

# dim>2数组相乘
a3 = np.expand_dims(a, axis=2)
PrintArray("a3", a3)
bt3 = np.expand_dims(b.T, axis=1)
PrintArray("bt3", bt3)
c3 = np.matmul(a3, bt3)
PrintArray("c3", c3)

# 当dim>2时， np.matmul与np.dot和经转换后的np.inner计算方式不同
# np.dot和经转换后的np.inner计算方式依旧相同
c3_dot = a3.dot(bt3)
PrintArray("c3_dot", c3_dot)
c3_inner = np.inner(a3, np.transpose(bt3, (0, 2, 1)))
PrintArray("c3_inner", c3_inner)

a:
 [[1 2 3]
 [4 5 6]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 3); 元素个数：6; 元素的字节大小：8
b:
 [[ 5  6]
 [ 7  8]
 [ 9 10]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 2); 元素个数：6; 元素的字节大小：8
c:
 [[ 46  52]
 [109 124]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
c_dot:
 [[ 46  52]
 [109 124]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
c_inner:
 [[ 46  52]
 [109 124]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
a3:
 [[[1]
  [2]
  [3]]

 [[4]
  [5]
  [6]]]
维度数(秩)：3; 元素类型：int64; 数组的维度形状：(2, 3, 1); 元素个数：6; 元素的字节大小：8
bt3:
 [[[ 5  7  9]]

 [[ 6  8 10]]]
维度数(秩)：3; 元素类型：int64; 数组的维度形状：(2, 1, 3); 元素个数：6; 元素的字节大小：8
c3:
 [[[ 5  7  9]
  [10 14 18]
  [15 21 27]]

 [[24 32 40]
  [30 40 50]
  [36 48 60]]]
维度数(秩)：3; 元素类型：int64; 数组的维度形状：(2, 3, 3); 元素个数：18; 元素的字节大小：8
c3_dot:
 [[[[ 5  7  9]
   [ 6  8 10]]

  [[10 14 18]
   [12 16 20]]

  [[15 21 27]
   [18 24 30]]]


 [[[20 28 36]
   [24 32 40]]

  [[25 35 45]
   [30 40 50]]

  [[30 42 54]
   [36 48 60]]]]
维度数(秩)：4; 元素类型：int64; 数组的维度形状：(2, 3,

* numpy.linalg.det() 函数计算输入矩阵的行列式，（方阵）。

In [41]:
a = np.random.randint(1, 10, size=(3, 3))
PrintArray("a", a)

a_det = np.linalg.det(a)
print("a_det:\n", round(a_det, 5), end=SplitLine)

# 可使用代数余子式展开计算
a11 = a[1:, 1:]
a12 = a[1:, ::2]
a13 = a[1:, :-1]
PrintArray("a11", a11)
PrintArray("a12", a12)
PrintArray("a13", a13)
a_det_ = a[0][0] * pow(-1, 1 + 1) * np.linalg.det(a11) + a[0][1] * pow(-1, 1 + 2) * np.linalg.det(a12) + a[0][2] * pow(-1, 1 + 3) * np.linalg.det(a13)
print("a_det_:\n", round(a_det_, 5), end=SplitLine)

a:
 [[3 5 3]
 [1 2 7]
 [8 8 5]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 3); 元素个数：9; 元素的字节大小：8
a_det:
 93.0
a11:
 [[2 7]
 [8 5]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
a12:
 [[1 7]
 [8 5]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
a13:
 [[1 2]
 [8 8]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(2, 2); 元素个数：4; 元素的字节大小：8
a_det_:
 93.0


* numpy.linalg.solve() 函数给出了矩阵形式的线性方程的解。

In [44]:
'''
考虑以下线性方程组：
x + y + z = 6
2y + 5z = -4
2x + 5y - z = 27
'''

A = np.array([[1, 1, 1], [0, 2, 5], [2, 5, -1]])
Y = np.array([[6,], [-4,], [27,]])
PrintArray("A", A)
PrintArray("Y", Y)

# 求解x, y, z
# X = A^(-1)Y
X = np.linalg.solve(A, Y)
PrintArray("X", X)

A:
 [[ 1  1  1]
 [ 0  2  5]
 [ 2  5 -1]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 3); 元素个数：9; 元素的字节大小：8
Y:
 [[ 6]
 [-4]
 [27]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 1); 元素个数：3; 元素的字节大小：8
X:
 [[ 5.]
 [ 3.]
 [-2.]]
维度数(秩)：2; 元素类型：float64; 数组的维度形状：(3, 1); 元素个数：3; 元素的字节大小：8


* numpy.linalg.inv() 函数计算矩阵的乘法逆矩阵。

In [45]:
PrintArray("A", A)
PrintArray("Y", Y)

X_ = np.linalg.inv(A).dot(Y)
PrintArray("X_", X_)

A:
 [[ 1  1  1]
 [ 0  2  5]
 [ 2  5 -1]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 3); 元素个数：9; 元素的字节大小：8
Y:
 [[ 6]
 [-4]
 [27]]
维度数(秩)：2; 元素类型：int64; 数组的维度形状：(3, 1); 元素个数：3; 元素的字节大小：8
X_:
 [[ 5.]
 [ 3.]
 [-2.]]
维度数(秩)：2; 元素类型：float64; 数组的维度形状：(3, 1); 元素个数：3; 元素的字节大小：8
